### PR TITLE
Iter5 - Переменные окружения

### PR DESCRIPTION
**Pull request by dcwk**

### CODE REVIEW

File: cmd/agent/main.go [hunk 1]
```diff
@@ -1,3 +1,115 @@
...
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"runtime"
+	"strconv"
+	"time"
+)
+
+var (
```
Comment (arttet) [inline cmd/agent/main.go]: CHANGE REQUEST: please don't use global variables

File: cmd/agent/main.go [hunk 2]
```diff
@@ -1,3 +1,115 @@
...
+
+	var pollCount int64 = 0
+
+	for {
+		time.Sleep(time.Duration(pollInterval) * time.Second)
+		gauges := getGauges()
+
+		if pollCount%reportInterval == 0 {
+			for k, v := range gauges {
+				r := bytes.NewReader([]byte(""))
```
Comment (arttet) [inline cmd/agent/main.go]: CHANGE REQUEST: please split your application by layers.

File: cmd/agent/main.go [hunk 3]
```diff
@@ -1,3 +1,115 @@
...
+	gauges["PauseTotalNs"] = float64(ms.PauseTotalNs)
+	gauges["StackInuse"] = float64(ms.StackInuse)
+	gauges["StackSys"] = float64(ms.StackSys)
+	gauges["Sys"] = float64(ms.Sys)
+	gauges["TotalAlloc"] = float64(ms.TotalAlloc)
+
+	return gauges
+}
+
+func parseFlags() {
```
Comment (arttet) [inline cmd/agent/main.go]: SUGGESTION: it would be better if it's in `flags.go`

File: cmd/agent/main.go [hunk 4]
```diff
@@ -1,3 +1,115 @@
...
+	flag.Parse()
+
+	if envAddress := os.Getenv("ADDRESS"); envAddress != "" {
+		flagServerAddr = envAddress
+	}
+
+	if envReportInterval := os.Getenv("REPORT_INTERVAL"); envReportInterval != "" {
+		interval, err := strconv.ParseInt(envReportInterval, 10, 64)
+		if err != nil {
+			log.Fatalln(err)
```
Comment (arttet) [inline cmd/agent/main.go]: CHANGE REQUEST: it's better if you return an error

**Review by arttet — CHANGES_REQUESTED** [review]
Please add
- https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- https://github.com/golangci/golangci-lint

File: cmd/agent/main.go (hunk 1 above)
Comment (dcwk) [reply cmd/agent/main.go]: \+

File: cmd/agent/main.go (hunk 3 above)
Comment (dcwk) [reply cmd/agent/main.go]: \+

File: cmd/agent/main.go (hunk 4 above)
Comment (dcwk) [reply cmd/agent/main.go]: \+

File: cmd/agent/main.go (hunk 2 above)
Comment (dcwk) [reply cmd/agent/main.go]: \+

File: internal/client/application.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"fmt"
+	"github.com/dcwk/metrics/internal/config"
+	"github.com/dcwk/metrics/internal/handlers"
+	"time"
+)
+
+func Run(conf config.ClientConf) error {
+	var pollCount int64 = 0
```
Comment (arttet) [inline internal/client/application.go]:
```suggestion
	var pollCount int64
```

File: internal/config/client.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package config
+
+import (
+	"flag"
+	"github.com/caarlos0/env"
```
Comment (arttet) [inline internal/config/client.go:6]: SUGGESTION: please sort imports like there

https://stackoverflow.com/questions/38704769/golang-import-grouping-by-package

File: internal/config/client.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	"log"
+)
+
+type ClientConf struct {
+	ServerAddr     string `env:"ADDRESS"`
+	ReportInterval int64  `env:"REPORT_INTERVAL"`
+	PollInterval   int64  `env:"POLL_INTERVAL"`
+}
+
+func NewClientConf() ClientConf {
```
Comment (arttet) [inline internal/config/client.go]:
```suggestion
func NewClientConf() *ClientConf, error {
```

File: internal/storage/storage.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
+package storage
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+var (
```
Comment (arttet) [inline internal/storage/storage.go]: CHANGE REQUEST: please avoid global variables

File: internal/handlers/updateMetric.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
+package handlers
+
+import (
+	"github.com/dcwk/metrics/internal/storage"
+	"github.com/go-chi/chi/v5"
+	"net/http"
+)
+
+func UpdateMetric(w http.ResponseWriter, r *http.Request) {
```
Comment (arttet) [inline internal/handlers/updateMetric.go]: SUGGESTION: you can use it as a method

File: internal/storage/storage.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	counterMx sync.RWMutex
+	counter   map[string]int64
+}
+
+type MemStorage struct {
+	Gauge
+	Counter
+}
+
+func GetStorage() *MemStorage {
```
Comment (arttet) [inline internal/storage/storage.go]: CHANGE REQUEST: unfortunately, I don't like singletons

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/storage/storage.go (hunk 10 above)
Comment (dcwk) [reply internal/storage/storage.go]: \+

File: internal/storage/storage.go (hunk 8 above)
Comment (dcwk) [reply internal/storage/storage.go]: \+

File: internal/config/client.go (hunk 7 above)
Comment (dcwk) [reply internal/config/client.go]: \+

File: internal/client/application.go (hunk 5 above)
Comment (dcwk) [reply internal/client/application.go]: \+

File: internal/config/client.go (hunk 6 above)
Comment (dcwk) [reply internal/config/client.go:6]: \+

File: internal/handlers/updateMetric.go (hunk 9 above)
Comment (dcwk) [reply internal/handlers/updateMetric.go]: не понял. Как использовать это как метод?

File: internal/config/client.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+type ClientConf struct {
+	ServerAddr     string `env:"ADDRESS"`
+	ReportInterval int64  `env:"REPORT_INTERVAL"`
+	PollInterval   int64  `env:"POLL_INTERVAL"`
+}
+
+func NewClientConf() (*ClientConf, error) {
+	conf := ClientConf{}
```
Comment (arttet) [inline internal/config/client.go]:
```suggestion
	conf := &ClientConf{}
```

File: internal/config/server.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+	"github.com/caarlos0/env"
+)
+
+type ServerConf struct {
+	ServerAddr string `env:"ADDRESS"`
+}
+
+func NewServerConf() (*ServerConf, error) {
+	conf := ServerConf{}
```
Comment (arttet) [inline internal/config/server.go]:
```suggestion
	conf := &ServerConf{}
```

File: internal/handlers/getAllMetrics.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+	"github.com/dcwk/metrics/internal/storage"
+)
+
+const (
+	gauge   = "gauge"
+	counter = "counter"
+)
+
+func GetAllMetrics(w http.ResponseWriter, r *http.Request, s storage.Storage) {
```
Comment (arttet) [inline internal/handlers/getAllMetrics.go]:
```suggestion
func (h *Handler) GetAllMetrics(w http.ResponseWriter, r *http.Request) {
```

File: internal/server/application.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	})
+
+	r.Route("/value/", func(r chi.Router) {
+		r.Get("/{type}/{name}", func(w http.ResponseWriter, r *http.Request) {
+			handlers.GetMetric(w, r, s)
+		})
+	})
+
+	r.Route("/update/", func(r chi.Router) {
+		r.Post("/{type}/{name}/{value}", func(w http.ResponseWriter, r *http.Request) {
```
Comment (arttet) [inline internal/server/application.go]: SUGGESTION: you can simplify it

File: internal/storage/storage.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type Storage interface {
+	AddGauge(name string, value string) error
```
Comment (arttet) [inline internal/storage/storage.go:11]: SUGGESTION:

https://go.dev/doc/effective_go

**Review by arttet — APPROVED** [review]
Great job!

File: internal/config/client.go (hunk 11 above)
Comment (dcwk) [reply internal/config/client.go]: \+

File: internal/config/server.go (hunk 12 above)
Comment (dcwk) [reply internal/config/server.go]: \+

File: internal/server/application.go (hunk 14 above)
Comment (dcwk) [reply internal/server/application.go]: \+

**Pull request merged**